File: autozig-math/src/mat3.rs
use autozig::include_zig;
use crate::{Vec3, Quat};

#[repr(C)]
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct Mat3 {
    pub cols: [[f32; 3]; 3],
}

include_zig!("zig/mat3.zig", {
    fn mat3_identity() -> Mat3;
    fn mat3_from_scale(scale: Vec3) -> Mat3;
    fn mat3_from_quat(q: Quat) -> Mat3;
    fn mat3_mul(self_: Mat3, other: Mat3) -> Mat3;
    fn mat3_mul_vec3(self_: Mat3, v: Vec3) -> Vec3;
    fn mat3_transpose(self_: Mat3) -> Mat3;
    fn mat3_determinant(self_: Mat3) -> f32;
    fn mat3_inverse(self_: Mat3) -> Mat3;
});

impl Mat3 {
    pub const IDENTITY: Self = Self {
        cols: [[1.0, 0.0, 0.0], [0.0, 1.0, 0.0], [0.0, 0.0, 1.0]],
    };

    pub const ZERO: Self = Self {
        cols: [[0.0, 0.0, 0.0], [0.0, 0.0, 0.0], [0.0, 0.0, 0.0]],
    };

    #[inline]
    pub fn identity() -> Self {
        mat3_identity()
    }

    /// Creates a 3x3 matrix from three column vectors.
    #[inline]
    pub fn from_cols(x_axis: Vec3, y_axis: Vec3, z_axis: Vec3) -> Self {
        Self {
            cols: [
                [x_axis.x, x_axis.y, x_axis.z],
                [y_axis.x, y_axis.y, y_axis.z],
                [z_axis.x, z_axis.y, z_axis.z],
            ],
        }
    }

    /// Creates a 3x3 matrix from a scale vector.
    #[inline]
    pub fn from_scale(scale: Vec3) -> Self {
        mat3_from_scale(scale)
    }

    /// Creates a 3x3 matrix from a quaternion.
    #[inline]
    pub fn from_quat(q: Quat) -> Self {
        mat3_from_quat(q)
    }

    /// Returns the first column (X axis).
    #[inline]
    pub fn x_axis(&self) -> Vec3 {
        Vec3::new(self.cols[0][0], self.cols[0][1], self.cols[0][2])
    }

    /// Returns the second column (Y axis).
    #[inline]
    pub fn y_axis(&self) -> Vec3 {
        Vec3::new(self.cols[1][0], self.cols[1][1], self.cols[1][2])
    }

    /// Returns the third column (Z axis).
    #[inline]
    pub fn z_axis(&self) -> Vec3 {
        Vec3::new(self.cols[2][0], self.cols[2][1], self.cols[2][2])
    }

    #[inline]
    pub fn mul_vec3(self, v: Vec3) -> Vec3 {
        mat3_mul_vec3(self, v)
    }

    #[inline]
    pub fn transpose(self) -> Self {
        mat3_transpose(self)
    }

    #[inline]
    pub fn determinant(self) -> f32 {
        mat3_determinant(self)
    }

    #[inline]
    pub fn inverse(self) -> Self {
        mat3_inverse(self)
    }
}

impl Default for Mat3 {
    fn default() -> Self {
        Self::IDENTITY
    }
}

impl std::ops::Mul for Mat3 {
    type Output = Self;
    fn mul(self, rhs: Self) -> Self::Output {
        mat3_mul(self, rhs)
    }
}

impl std::ops::Mul<Vec3> for Mat3 {
    type Output = Vec3;
    fn mul(self, rhs: Vec3) -> Self::Output {
        mat3_mul_vec3(self, rhs)
    }
}

File: autozig-math/zig/mat3.zig
const std = @import("std");
const Vec3 = @import("vec3.zig").Vec3;
const Quat = @import("quat.zig").Quat;

pub const Mat3 = extern struct {
    cols: [3][3]f32,

    pub const IDENTITY = Mat3{ .cols = .{
        .{ 1.0, 0.0, 0.0 },
        .{ 0.0, 1.0, 0.0 },
        .{ 0.0, 0.0, 1.0 },
    } };

    pub const ZERO = Mat3{ .cols = .{
        .{ 0.0, 0.0, 0.0 },
        .{ 0.0, 0.0, 0.0 },
        .{ 0.0, 0.0, 0.0 },
    } };

    pub fn identity() Mat3 {
        return IDENTITY;
    }

    pub fn from_cols(c0: Vec3, c1: Vec3, c2: Vec3) Mat3 {
        return .{ .cols = .{
            .{ c0.x, c0.y, c0.z },
            .{ c1.x, c1.y, c1.z },
            .{ c2.x, c2.y, c2.z },
        } };
    }

    pub fn from_scale(scale: Vec3) Mat3 {
        return .{ .cols = .{
            .{ scale.x, 0.0, 0.0 },
            .{ 0.0, scale.y, 0.0 },
            .{ 0.0, 0.0, scale.z },
        } };
    }

    pub fn from_quat(q: Quat) Mat3 {
        const x2 = q.x + q.x;
        const y2 = q.y + q.y;
        const z2 = q.z + q.z;
        const xx = q.x * x2;
        const xy = q.x * y2;
        const xz = q.x * z2;
        const yy = q.y * y2;
        const yz = q.y * z2;
        const zz = q.z * z2;
        const wx = q.w * x2;
        const wy = q.w * y2;
        const wz = q.w * z2;

        return .{ .cols = .{
            .{ 1.0 - (yy + zz), xy + wz, xz - wy },
            .{ xy - wz, 1.0 - (xx + zz), yz + wx },
            .{ xz + wy, yz - wx, 1.0 - (xx + yy) },
        } };
    }

    pub fn mul(self: Mat3, other: Mat3) Mat3 {
        var result: Mat3 = undefined;
        inline for (0..3) |col| {
            inline for (0..3) |row| {
                var sum: f32 = 0.0;
                inline for (0..3) |k| {
                    sum += self.cols[k][row] * other.cols[col][k];
                }
                result.cols[col][row] = sum;
            }
        }
        return result;
    }

    pub fn mul_vec3(self: Mat3, v: Vec3) Vec3 {
        return .{
            .x = self.cols[0][0] * v.x + self.cols[1][0] * v.y + self.cols[2][0] * v.z,
            .y = self.cols[0][1] * v.x + self.cols[1][1] * v.y + self.cols[2][1] * v.z,
            .z = self.cols[0][2] * v.x + self.cols[1][2] * v.y + self.cols[2][2] * v.z,
        };
    }

    pub fn transpose(self: Mat3) Mat3 {
        return .{ .cols = .{
            .{ self.cols[0][0], self.cols[1][0], self.cols[2][0] },
            .{ self.cols[0][1], self.cols[1][1], self.cols[2][1] },
            .{ self.cols[0][2], self.cols[1][2], self.cols[2][2] },
        } };
    }

    pub fn determinant(self: Mat3) f32 {
        const a = self.cols[0][0];
        const b = self.cols[1][0];
        const c = self.cols[2][0];
        const d = self.cols[0][1];
        const e = self.cols[1][1];
        const f = self.cols[2][1];
        const g = self.cols[0][2];
        const h = self.cols[1][2];
        const i = self.cols[2][2];
        return a * (e * i - f * h) - b * (d * i - f * g) + c * (d * h - e * g);
    }

    pub fn inverse(self: Mat3) Mat3 {
        const det = self.determinant();
        if (@abs(det) < std.math.floatEps(f32)) {
            return IDENTITY;
        }

        const a = self.cols[0][0];
        const b = self.cols[1][0];
        const c = self.cols[2][0];
        const d = self.cols[0][1];
        const e = self.cols[1][1];
        const f = self.cols[2][1];
        const g = self.cols[0][2];
        const h = self.cols[1][2];
        const i = self.cols[2][2];

        const inv_det = 1.0 / det;

        return .{ .cols = .{
            .{ (e * i - f * h) * inv_det, (c * h - b * i) * inv_det, (b * f - c * e) * inv_det },
            .{ (f * g - d * i) * inv_det, (a * i - c * g) * inv_det, (c * d - a * f) * inv_det },
            .{ (d * h - e * g) * inv_det, (b * g - a * h) * inv_det, (a * e - b * d) * inv_det },
        } };
    }
};

export fn mat3_identity() Mat3 {
    return Mat3.identity();
}

export fn mat3_from_scale(scale: Vec3) Mat3 {
    return Mat3.from_scale(scale);
}

export fn mat3_from_quat(q: Quat) Mat3 {
    return Mat3.from_quat(q);
}

export fn mat3_mul(self: Mat3, other: Mat3) Mat3 {
    return self.mul(other);
}

export fn mat3_mul_vec3(self: Mat3, v: Vec3) Vec3 {
    return self.mul_vec3(v);
}

export fn mat3_transpose(self: Mat3) Mat3 {
    return self.transpose();
}

export fn mat3_determinant(self: Mat3) f32 {
    return self.determinant();
}

export fn mat3_inverse(self: Mat3) Mat3 {
    return self.inverse();
}
